### PR TITLE
[#770] Replace agency ids with slugs

### LINKED ISSUE #770: [Agency Dashboard] Change URL from </agency/<agency_id> to </agency-slug>
Agency slug is a lowercase, hyphen separated version of agency name. This is blocked by having a backend endpoint that accepts the agency slug instead of the agency id, because otherwise the FE has no way of converting the slug to an ID to hit the existing endpoint.

Note: PR with backend endpoint is [here](https://github.com/Recidiviz/recidiviz-data/pull/22258)

### PR DESCRIPTION
**Pull request by alfonsogoberjr**
## Description of the change

Replace `/agency/:id` navigation structure with `/agency/:slug`
Replace calls to `/agency/:id/published_data` with `/agency/:slug/published_data`

## Type of change
Feature

## Related issues

Closes #770 

### Development

**This box MUST be checked by the submitter prior to merging**:
- [x] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [x] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: agency-dashboard/src/DashboardView/DashboardView.tsx [hunk 1]
```diff
@@ -109,7 +112,7 @@ export const DashboardView = observer(() => {
     useState<boolean>(false);
   const navigate = useNavigate();
   const params = useParams();
-  const agencyId = Number(params.id);
+  const agencySlug = slugify(params.slug as string);
```
Comment (alfonsogoberjr) [inline agency-dashboard/src/DashboardView/DashboardView.tsx:115]: Need to change this

File: agency-dashboard/src/AgencyOverview/AgencyOverview.tsx [hunk 2]
```diff
@@ -70,69 +74,114 @@ const orderedCategoriesMap: {
 };
 
 // include here systems that you want users to be able to choose
-const availableSystems: AgencySystems[] = ["PRISONS"];
+const availableSystems: AgencySystems[] = ["PRISONS", "LAW_ENFORCEMENT"];
```
Comment (lilidworkin) [inline agency-dashboard/src/AgencyOverview/AgencyOverview.tsx:77]: Why is Law Enforcement added?

File: agency-dashboard/src/AgencyOverview/AgencyOverview.tsx [hunk 3]
```diff
@@ -70,69 +74,114 @@ const orderedCategoriesMap: {
...
 
 export const AgencyOverview = observer(() => {
   const navigate = useNavigate();
   const params = useParams();
-  const agencyId = Number(params.id);
+  const [agencySlug, setAgencySlug] = useState<string | undefined>(params.slug);
```
Comment (lilidworkin) [inline agency-dashboard/src/AgencyOverview/AgencyOverview.tsx:82]: Probably my FE ignorance talking, but I'm curious why we're using useState here, when before we just referenced the params directly. Is there a reason we can't do `const agencySlug = params.slug`?

File: agency-dashboard/src/AgencyOverview/AgencyOverview.tsx [hunk 4]
```diff
@@ -70,69 +74,114 @@ const orderedCategoriesMap: {
...
         });
       }
-    };
-    fetchData();
+    }
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
 
+  useEffect(() => {
+    console.log(agencyDataStore.agency, availableSystems);
```
Comment (lilidworkin) [inline agency-dashboard/src/AgencyOverview/AgencyOverview.tsx:139]: Did you mean to commit this?

File: agency-dashboard/src/AgencyOverview/AgencyOverview.tsx [hunk 5]
```diff
@@ -70,69 +74,114 @@ const orderedCategoriesMap: {
...
           timeout: 4000,
         });
       }
-    };
-    fetchData();
+    }
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
 
+  useEffect(() => {
```
Comment (lilidworkin) [inline agency-dashboard/src/AgencyOverview/AgencyOverview.tsx:138]: What's the reasoning for these changes? Seems unrelated to the slug changes.

File: agency-dashboard/src/DashboardView/DashboardView.tsx [hunk 6]
```diff
@@ -150,32 +158,36 @@ export const DashboardView = observer(() => {
...
       !agencyDataStore.loading &&
       !agencyDataStore.dimensionNamesByMetricAndDisaggregation[metricKeyParam]
     ) {
-      navigate(`/agency/${agencyId}`);
+      navigate(`/agency/${encodeURIComponent(agencySlug)}`);
     }
 
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [agencyDataStore.loading]);
+  }, [navigate, agencyDataStore.loading, agencySlug]);
```
Comment (lilidworkin) [inline agency-dashboard/src/DashboardView/DashboardView.tsx:190]: OOC why does navigate need to be a dependency?

**Review by lilidworkin — COMMENTED** [review]
Looks great to me! @mxosman should take a look too though. 

@alfonsogoberjr could you either upload a loom video showing that this works, or deploy to a playtesting branch so we can test it out? TY!

File: agency-dashboard/src/AgencyOverview/AgencyOverview.tsx (hunk 5 above)
Comment (alfonsogoberjr) [reply agency-dashboard/src/AgencyOverview/AgencyOverview.tsx:139]: A lot of issues were caused by lack of React dependency handling. A component's scope is not static; it is actually re-rendered constantly and the default is to re-assign/re-allocate etc. any logic not tied to a useEffect/useCallback/useMemo dependency array. As you know, an empty array guarantees the scope will only be rendered once, on document load. 

useEffect is for scope that you don't care about returning values from
useCallback is for scope that you want a function as a return value from  
useMemo is for scope that you want to return anything else (primitives and objects) as a return value from 

There are a lot of assignments and object transformations done outside of these contexts, putting them in an unpredictable state and keeping them out of sync with variables whose value they depend on.

File: agency-dashboard/src/DashboardView/DashboardView.tsx (hunk 6 above)
Comment (alfonsogoberjr) [reply agency-dashboard/src/DashboardView/DashboardView.tsx:190]: Functions are first-class citizens in Javascript/Typescript, and their values are not guaranteed to be assigned in the scope of a useEffect/useCallback/useMemo. Every dependency variable has to be checked for existence before you can do anything to it, as usually the first render has undefined values. I don't know why that is, though

Comment (alfonsogoberjr) [reply agency-dashboard/src/DashboardView/DashboardView.tsx:190]: It's actually also why we shouldn't be doing `// eslint-disable-next-line react-hooks/exhaustive-deps` as much as we do

File: agency-dashboard/src/AgencyOverview/AgencyOverview.tsx (hunk 4 above)
Comment (alfonsogoberjr) [reply agency-dashboard/src/AgencyOverview/AgencyOverview.tsx]: lol nope, removed

File: agency-dashboard/src/AgencyOverview/AgencyOverview.tsx (hunk 3 above)
Comment (alfonsogoberjr) [reply agency-dashboard/src/AgencyOverview/AgencyOverview.tsx:82]: See [here](https://github.com/Recidiviz/justice-counts/pull/787#discussion_r1275486680) and [here](https://github.com/Recidiviz/justice-counts/pull/787#discussion_r1275488172)

File: agency-dashboard/src/AgencyOverview/AgencyOverview.tsx (hunk 2 above)
Comment (alfonsogoberjr) [reply agency-dashboard/src/AgencyOverview/AgencyOverview.tsx:77]: I was testing data that I had locally, will revert this

File: agency-dashboard/src/DashboardView/DashboardView.tsx [hunk 7]
```diff
@@ -150,32 +158,36 @@ export const DashboardView = observer(() => {
...
+            color: "red",
+            timeout: 4000,
+          });
+        }
+      };
+      fetchData();
+    }
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, []);
+  }, [agencySlug]);
```
Comment (mxosman) [inline agency-dashboard/src/DashboardView/DashboardView.tsx:176]: Should we add `agencyDataStore` as a dependency here to reduce the `// eslint-disable-next-line react-hooks/exhaustive-deps`?

File: agency-dashboard/src/DashboardView/DashboardView.tsx [hunk 8]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable simple-import-sort/imports */
```
Comment (mxosman) [inline agency-dashboard/src/DashboardView/DashboardView.tsx:1]: Is this for the `bluebird` library? Doesn't have to be in this PR - but maybe it's worth updating our linter rules at some point.

**Review by mxosman — COMMENTED** [review]
This is super cool, Alfonso! Left a bunch of questions and comments - mostly for my learnings and to better understand your perspective.

File: agency-dashboard/src/DashboardView/DashboardView.tsx (hunk 6 above)
Comment (alfonsogoberjr) [reply agency-dashboard/src/DashboardView/DashboardView.tsx:190]: yes I'm doing exactly that haha

**Comment by alfonsogoberjr** [comment]
Playtesting https://alfonsotest---agency-dashboard-web-b47yvyxs3q-uc.a.run.app

File: agency-dashboard/src/DashboardView/DashboardView.tsx [hunk 9]
```diff
@@ -228,28 +240,58 @@ export const DashboardView = observer(() => {
...
+    if (
+      agencyDataStore.agency &&
+      agencyDataStore.metricsByKey?.[metricKeyParam] &&
+      metricKeyParam
+    ) {
+      const metric = agencyDataStore.metricsByKey[metricKeyParam];
+      if (metric) {
+        each(
+          metric.filenames,
+          downloadFeedData(metric.system.key, agencyDataStore.agency.id)
```
[the remark this answers is not in the document]
Comment (alfonsogoberjr) [reply agency-dashboard/src/DashboardView/DashboardView.tsx:262]: Yep

File: agency-dashboard/src/AgencyOverview/AgencyOverview.tsx [hunk 10]
```diff
@@ -75,63 +79,136 @@ const availableSystems: AgencySystems[] = ["PRISONS"];
...
+  const [
+    metricsByAvailableCategoriesAndSystemsWithData,
+    setMetricsByAvailableCategoriesAndSystemsWithData,
+  ] = useState<Metric[]>();
+  const [agencyDescription, setAgencyDescription] = useState<string>(
+    agencyDataStore.agencySettingsBySettingType.PURPOSE_AND_FUNCTIONS?.value
+  );
+  const [agencyHomepageUrl, setAgencyHomepageUrl] = useState<string>(
+    agencyDataStore.agencySettingsBySettingType.HOMEPAGE_URL?.value
+  );
```
[the remark this answers is not in the document]
Comment (alfonsogoberjr) [reply agency-dashboard/src/AgencyOverview/AgencyOverview.tsx:104]: Yes I agree! It would be a bigger lift than the scope of this PR however

File: agency-dashboard/src/DashboardView/DashboardView.tsx (hunk 8 above)
Comment (alfonsogoberjr) [reply agency-dashboard/src/DashboardView/DashboardView.tsx:1]: Yes I agree

**Comment by alfonsogoberjr** [comment]
by the way I'm putting API documentation in a follow-up PR just so this can get merged. The React rendering issues I've been fixing are just piecemeal because refactoring the whole codebase is out of scope currently.

File: agency-dashboard/src/DashboardView/DashboardView.tsx [hunk 11]
```diff
@@ -149,33 +173,37 @@ export const DashboardView = observer(() => {
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
 
-  useEffect(() => {
-    const fetchData = async () => {
+  useAsyncEffect(async () => {
```
Comment (mxosman) [inline agency-dashboard/src/DashboardView/DashboardView.tsx:176]: Haha - nice! I didn't know this package existed!

File: agency-dashboard/src/AgencyOverview/AgencyOverview.tsx [hunk 12]
```diff
@@ -14,6 +14,23 @@
...
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+// =============================================================================
```
Comment (mxosman) [inline agency-dashboard/src/AgencyOverview/AgencyOverview.tsx:33]: Duplicate header

File: agency-dashboard/src/stores/AgencyDataStore.ts [hunk 13]
```diff
@@ -156,13 +156,15 @@ class AgencyDataStore {
...
 
-  async fetchAgencyData(agencyId: number): Promise<void | Error> {
+  async fetchAgencyData(agencySlug: string): Promise<void | Error> {
     try {
       runInAction(() => {
         this.loading = true;
       });
       const response = (await request({
-        path: `/api/agencies/${agencyId}/published_data`,
+        path: `/api/v2/agencies/${encodeURIComponent(
```
[the remark this answers is not in the document]
Comment (mxosman) [reply agency-dashboard/src/stores/AgencyDataStore.ts:165]: Ah - I see it more clearly when going to an agency with [DEMO] in its name (e.g. `https://alfonsotest---agency-dashboard-web-b47yvyxs3q-uc.a.run.app/agency/prisons-%5Bdemo%5D`) - are we OK with the user-facing URL having the encoding visible? I don't see any harm - just mostly aesthetics as both seem to work just fine (e.g. `/agency/prisons-[demo]` and `/agency/prisons-%5Bdemo%5D`).

**Review by mxosman — APPROVED** [review]
None of my feedback is blocking! This works great - thank you for entertaining my questions and helping instill better practices!

File: agency-dashboard/src/stores/AgencyDataStore.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (alfonsogoberjr) [reply agency-dashboard/src/stores/AgencyDataStore.ts:165]: Yes exactly. I think we can change the user-facing url however to the non-encoded value. I'll add that in my documentation PR

File: agency-dashboard/src/AgencyOverview/AgencyOverview.tsx [hunk 14]
```diff
@@ -75,63 +79,136 @@ const availableSystems: AgencySystems[] = ["PRISONS"];
...
+    agencyDataStore.agencySettingsBySettingType.HOMEPAGE_URL?.value
+  );
 
-  const handleNavigate = (isPublished: boolean, metricKey: string) => {
-    if (isPublished) {
-      navigate(
-        `/agency/${agencyId}/dashboard?metric=${metricKey.toLocaleLowerCase()}`
-      );
+  useEffect(() => {
+    if (params.slug) {
```
[the remark this answers is not in the document]
Comment (alfonsogoberjr) [reply agency-dashboard/src/AgencyOverview/AgencyOverview.tsx]: Yes because it is initially undefined on the first (sometimes two) renders. I think our navigation library needs a couple of renders to activate with the url param value

File: agency-dashboard/src/stores/AgencyDataStore.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (lilidworkin) [reply agency-dashboard/src/stores/AgencyDataStore.ts:165]: @alfonsogoberjr let's hold off on the documentation PR for now and revisit it after CLA -- I think it's higher priority to start on the line graph tasks
